packages/xdl/src/Project.js

```diff
@@ -2139,6 +2142,163 @@ export async function getUrlAsync(projectRoot: string, options: Object = {}) {
+    assetInfo[newHash] = true;
+
+    if (options.save) {
+      if (hash === newHash) {
+        logger.global.info(
+          chalk.gray(
+            `Compressed asset ${image} is identical to the original. Using original instead.`
+          )
+        );
+        fs.unlinkSync(newName);
+      } else {
+        logger.global.info(chalk.gray(`Saving original asset to ${newName}`));
+        // Save the old hash to prevent reoptimizing
+        assetInfo[hash] = true;
+      }
+    } else {
+      // Delete the renamed original asset
+      fs.unlinkSync(newName);
+    }
+    totalSaved += amountSaved;
+    logger.global.info(chalk.yellow(`Saved ${toReadableValue(amountSaved)}`));
+  }
+  if (totalSaved === 0) {
+    logger.global.info('No assets optimized. Everything is fully compressed!');
+  } else {
+    logger.global.info(
+      `Finished compressing assets. ${chalk.green(toReadableValue(totalSaved))} saved.`
+    );
+  }
+  // Remove assets that have been deleted/modified from assets.json
+  outdated.forEach(outdatedHash => {
+    delete assetInfo[outdatedHash];
+  });
+  assetJson.writeAsync(assetInfo);
+}
+
+const toReadableValue = bytes => {
+  const sizes = ['Bytes', 'KB', 'MB', 'GB'];
+  const index = Math.floor(Math.log(bytes) / Math.log(1024));
+  const reduced = (bytes / Math.pow(1024, index)).toFixed(2) * 1;
+
+  return `${reduced} ${sizes[index]}`;
+};
+
+/*
+ * Calculate SHA256 Checksum value of a file based on its contents
+ */
+const calculateHash = file => {
+  const contents = fs.readFileSync(file);
+  return crypto
+    .createHash('sha256')
+    .update(contents)
+    .digest('hex');
+};
+
+/*
+ * Compress an inputted jpg or png and save original copy with .expo extension
+ */
+const optimizeImage = async (image, newName) => {
+  logger.global.info(`Optimizing ${image}`);
+  // Rename the file with .expo extension
+  fs.copyFileSync(image, newName);
+
+  // Extract the format and compress
+  const buffer = await sharp(image).toBuffer();
+  const { format } = await sharp(buffer).metadata();
+  if (format === 'jpeg') {
+    await sharp(newName)
+      .jpeg({ quality: 60 })
+      .toFile(image)
+      .catch(err => logger.global.error(err));
+  } else {
+    await sharp(newName)
+      .png({ quality: 60 })
+      .toFile(image)
+      .catch(err => logger.global.error(err));
+  }
+};
+
+/*
+ * Find all project assets under assetBundlePatterns in app.json excluding node_modules
+ */
+const getAssetFiles = (exp, projectDir) => {
+  const { assetBundlePatterns } = exp;
+  const options = { cwd: projectDir, ignore: '**/node_modules/**' };
+  const files = [];
+  assetBundlePatterns.forEach(pattern => {
+    files.push(...glob.sync(pattern, options));
+  });
+  return files.map(file => `${projectDir}/${file}`.replace('//', '/'));
+};
+
+/*
+ * Read the contents of assets.json under .expo-shared folder. Create the file/directory if they don't exist.
+ */
+const readAssetJsonAsync = async projectDir => {
```

Comment:
i would suggest moving this (and possibly some of these other helper functions) to `ProjectUtils` or perhaps even a new `AssetUtils` module, since this file is already gigantic

Reply:
Yeah, I was thinking about that too but wasn't sure where to put them. `AssetUtils` sounds good!